src/meta/processors/indexMan/RebuildEdgeIndexProcessor.cpp

```diff
@@ -0,0 +1,125 @@
+/* Copyright (c) 2019 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "meta/ActiveHostsMan.h"
+#include "meta/processors/admin/AdminClient.h"
+#include "meta/processors/indexMan/RebuildEdgeIndexProcessor.h"
+
+DECLARE_int32(heartbeat_interval_secs);
+
+namespace nebula {
+namespace meta {
+
+void RebuildEdgeIndexProcessor::process(const cpp2::RebuildIndexReq& req) {
+    auto space = req.get_space_id();
+    CHECK_SPACE_ID_AND_RETURN(space);
+    const auto &indexName = req.get_index_name();
+    auto isOffline = req.get_is_offline();
+    if (!isOffline) {
+        LOG(ERROR) << "Currently not support online rebuild index";
+        resp_.set_code(cpp2::ErrorCode::E_UNSUPPORTED);
+        onFinished();
+        return;
+    }
+
+    LOG(INFO) << "Rebuild Edge Index Space " << space << ", Index Name " << indexName;
+    const auto& hostPrefix = MetaServiceUtils::leaderPrefix();
+    std::unique_ptr<kvstore::KVIterator> leaderIter;
+    auto kvRet = kvstore_->prefix(kDefaultSpaceId, kDefaultPartId, hostPrefix, &leaderIter);
+    if (kvRet != kvstore::ResultCode::SUCCEEDED) {
+        LOG(ERROR) << "Get space " << space << "'s part failed";
+        resp_.set_code(cpp2::ErrorCode::E_NOT_FOUND);
+        onFinished();
+        return;
+    }
+
+    folly::SharedMutex::ReadHolder indexHolder(LockUtils::edgeIndexLock());
+    auto edgeIndexIDResult = getIndexID(space, indexName);
+    if (!edgeIndexIDResult.ok()) {
+        LOG(ERROR) << "Edge Index " << indexName << " Not Found";
+        resp_.set_code(cpp2::ErrorCode::E_NOT_FOUND);
+        onFinished();
+        return;
+    }
+
+    auto edgeIndexID = edgeIndexIDResult.value();
+    auto edgeKey = MetaServiceUtils::indexKey(space, edgeIndexIDResult.value());
+    auto edgeResult = doGet(edgeKey);
+    if (!edgeResult.ok()) {
+        resp_.set_code(cpp2::ErrorCode::E_NOT_FOUND);
+        onFinished();
+        return;
+    }
+
+    auto statusKey = MetaServiceUtils::rebuildIndexStatus(space, 'E', indexName);
+    if (!MetaCommon::saveRebuildStatus(kvstore_, statusKey, "RUNNING")) {
```

Comment:
How about read statusKey key first, if it is running, just reject following request?

Reply:
Sure But when the cluster crashed and you want to restart this building processing. It maybe a problem.